specs/_features/verge/beacon-chain.md

```diff
@@ -0,0 +1,208 @@
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+<!-- /TOC -->
+
+## Introduction
+
+This upgrade adds transaction execution to the beacon chain as part of the Verge upgrade.
+
+## Custom types
+
+| Name | SSZ equivalent | Description |
+| - | - | - |
+| `StateDiff` | `List[StemStateDiff, MAX_STEMS]` | Only valid if list is sorted by stems |
+| `BanderwagonGroupElement` | `Bytes32` | |
+| `BanderwagonFieldElement` | `Bytes32` | |
+| `Stem` | `Bytes31` | |
+
+## Preset
+
+### Execution
+
+| Name | Value |
+| - | - |
+| `MAX_STEMS` | `2**16` |
+| `MAX_COMMITMENTS_PER_STEM` | `33` |
+| `VERKLE_WIDTH` | `256` |
+| `IPA_PROOF_DEPTH` | `8` |
+
+## Containers
+
+### Extended containers
+
+#### `ExecutionPayload`
+
+```python
+class ExecutionPayload(Container):
+    # Execution block header fields
+    parent_hash: Hash32
+    fee_recipient: ExecutionAddress  # 'beneficiary' in the yellow paper
+    state_root: Bytes32
+    receipts_root: Bytes32
+    logs_bloom: ByteVector[BYTES_PER_LOGS_BLOOM]
+    prev_randao: Bytes32  # 'difficulty' in the yellow paper
+    block_number: uint64  # 'number' in the yellow paper
+    gas_limit: uint64
+    gas_used: uint64
+    timestamp: uint64
+    extra_data: ByteList[MAX_EXTRA_DATA_BYTES]
+    base_fee_per_gas: uint256
+    block_hash: Hash32  # Hash of execution block
+    # Extra payload field
+    transactions: List[Transaction, MAX_TRANSACTIONS_PER_PAYLOAD]
+    execution_witness: ExecutionWitness  # [New in Verge]
+```
+
+#### `ExecutionPayloadHeader`
+
+```python
+class ExecutionPayloadHeader(Container):
+    # Execution block header fields
+    parent_hash: Hash32
+    fee_recipient: ExecutionAddress
+    state_root: Bytes32
+    receipts_root: Bytes32
+    logs_bloom: ByteVector[BYTES_PER_LOGS_BLOOM]
+    prev_randao: Bytes32
+    block_number: uint64
+    gas_limit: uint64
+    gas_used: uint64
+    timestamp: uint64
+    extra_data: ByteList[MAX_EXTRA_DATA_BYTES]
+    base_fee_per_gas: uint256
+    block_hash: Hash32  # Hash of execution block
+    transactions_root: Root
+    # Extra payload fields
+    execution_witness: ExecutionWitness
+```
+
+### New containers
+
+#### `SuffixStateDiff`
+
+```python
+class SuffixStateDiff(Container):
+    suffix: Byte
+
+    # Null means not currently present
+    current_value: Optional[Bytes32]
+
+    # Null means value not updated
+    new_value: Optional[Bytes32]
+```
+
+*Note*: on the Kaustinen testnet, `new_value` is ommitted from the container.
+
+#### `StemStateDiff`
+
+```python
+class StemStateDiff(Container):
+    stem: Stem
+    # Valid only if list is sorted by suffixes
+    suffix_diffs: List[SuffixStateDiff, VERKLE_WIDTH]
+```
+
+```python
+# Valid only if list is sorted by stems
+StateDiff = List[StemStateDiff, MAX_STEMS]
+```
+
+#### `IPAProof`
+
+```python
+class IpaProof(Container):
+    C_L = Vector[BanderwagonGroupElement, IPA_PROOF_DEPTH]
+    C_R = Vector[BanderwagonGroupElement, IPA_PROOF_DEPTH]
```

Comment:
The casing here is kind of "non standard". AFAIK Teku and Lodestar convert casing from snake_case to camelCase for all spec properties. `C_L`, `C_R` and `D` (in VerkleProof) do not really conform to either casing. @gballet could naming be reviewed for those?

Reply:
This isn't meant to be read in snake case or camel case, but as LaTeX code. I guess we could rename them `c_l` and `c_d` but let's discuss that during an ACDC just to make sure.

Reply:
using all lowercase for those fields is good :+1: